fibheap/fibheap.go

```diff
@@ -0,0 +1,384 @@
+	return heap.size
+}
+
+// DequeueMin removes and returns the
+// minimal element in the heap
+func (heap *FloatingFibonacciHeap) DequeueMin() (*Entry, error) {
+	if heap.IsEmpty() {
+		return nil, fmt.Errorf("Cannot dequeue minimum of empty heap")
+	}
+
+	heap.size--
+
+	// Copy pointer. Will need it later.
+	min := heap.min
+
+	if min.next == min { // This is the only root node
+		heap.min = nil
+	} else { // There are more root nodes
+		heap.min.prev.next = heap.min.next
+		heap.min.next.prev = heap.min.prev
+		heap.min = heap.min.next // Arbitrary element of the root list
+	}
+
+	if min.child != nil {
+		// Keep track of the first visited node
+		curr := min.child
+		for ok := true; ok; ok = (curr != min.child) {
+			curr.parent = nil
+			curr = curr.next
+		}
+	}
+
+	heap.min = mergeLists(heap.min, min.child)
+
+	if heap.min == nil {
+		// If there are no entries left, we're done.
+		return min, nil
+	}
+
+	treeSlice := make([]*Entry, 0, heap.size)
+	toVisit := make([]*Entry, 0, heap.size)
+
+	for curr := heap.min; len(toVisit) == 0 || toVisit[0] != curr; curr = curr.next {
+		toVisit = append(toVisit, curr)
+	}
+
+	for _, curr := range toVisit {
+		for {
+			for curr.degree >= len(treeSlice) {
+				treeSlice = append(treeSlice, nil)
+			}
+
+			if treeSlice[curr.degree] == nil {
+				treeSlice[curr.degree] = curr
+				break
+			}
+
+			other := treeSlice[curr.degree]
+			treeSlice[curr.degree] = nil
+
+			// Determine which of two trees has the smaller root
+			var minT, maxT *Entry
+			if other.Priority < curr.Priority {
+				minT = other
+				maxT = curr
+			} else {
+				minT = curr
+				maxT = other
+			}
+
+			// Break max out of the root list,
+			// then merge it into min's child list
+			maxT.next.prev = maxT.prev
+			maxT.prev.next = maxT.next
+
+			// Make it a singleton so that we can merge it
+			maxT.prev = maxT
+			maxT.next = maxT
+			minT.child = mergeLists(minT.child, maxT)
+
+			// Reparent max appropriately
+			maxT.parent = minT
+
+			// Clear max's mark, since it can now lose another child
+			maxT.marked = false
+
+			// Increase min's degree. It has another child.
+			minT.degree++
+
+			// Continue merging this tree
+			curr = minT
+		}
+
+		/* Update the global min based on this node.  Note that we compare
+		 * for <= instead of < here.  That's because if we just did a
+		 * reparent operation that merged two different trees of equal
+		 * priority, we need to make sure that the min pointer points to
+		 * the root-level one.
+		 */
+		if curr.Priority <= heap.min.Priority {
+			heap.min = curr
+		}
+	}
+
+	// All done. Return minimum element and no error
+	return min, nil
+}
+
+// DecreaseKey decreases the key of the
+// given element, sets it to the new
+// given priority and returns the node
+// if successfully set
+func (heap *FloatingFibonacciHeap) DecreaseKey(node *Entry, newPriority float64) (*Entry, error) {
+
+	if heap.IsEmpty() {
+		return nil, fmt.Errorf("Cannot decrease key in an empty heap")
```

Comment:
You might consider returning typed errors here.  It gives the caller an easy way to differentiate between errors.